Headers/DebugServer2/Utils/Bits.h

```diff
@@ -0,0 +1,88 @@
+//
+// Copyright (c) 2014-present, Facebook, Inc.
+// All rights reserved.
+//
+// This source code is licensed under the University of Illinois/NCSA Open
+// Source License found in the LICENSE file in the root directory of this
+// source tree. An additional grant of patent rights can be found in the
+// PATENTS file in the same directory.
+//
+
+#ifndef __DebugServer2_Utils_Bits_h
+#define __DebugServer2_Utils_Bits_h
+
+#include <type_traits>
+
+#include "DebugServer2/Utils/CompilerSupport.h"
+
+namespace ds2 {
+namespace Utils {
+namespace {
+// We apply the builtin bit functions on chunks of sizeof(int) * 8 bits.
+// However, we can't do full shifts of this size because only the least
+// signifitive bits are used and clang complains, so we shift twice half
+// the size.
+const size_t kHalfIntBits = sizeof(int) * 4;
```

Comment:
> *An unnamed namespace or a namespace declared directly or indirectly
within an unnamed namespace has internal linkage*.

I hadn't seen that `namespace {}` was enclosing this. So yeah, `static` is
useless in this case.

On Thu, Mar 10, 2016 at 11:44 AM, Dave Lee <notifications@github.com> wrote:

> In Headers/DebugServer2/Utils/Bits.h
> <https://github.com/facebook/ds2/pull/211#discussion_r55738070>:
>
> > +
> > +#ifndef __DebugServer2_Utils_Bits_h
> > +#define __DebugServer2_Utils_Bits_h
> > +
> > +#include <type_traits>
> > +
> > +#include "DebugServer2/Utils/CompilerSupport.h"
> > +
> > +namespace ds2 {
> > +namespace Utils {
> > +namespace {
> > +// We apply the builtin bit functions on chunks of sizeof(int) * 8 bits.
> > +// However, we can't do full shifts of this size because only the least
> > +// signifitive bits are used and clang complains, so we shift twice half
> > +// the size.
> > +const size_t kHalfIntBits = sizeof(int) * 4;
>
> According to cppreference.com
> <http://en.cppreference.com/w/cpp/language/storage_duration>
>
> Any of the following names declared at namespace scope have internal
> linkage
> …
>
>    - non-volatile const-qualified variables (including constexpr) that
>    aren't declared extern and aren't previously declared to have external
>    linkage.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/facebook/ds2/pull/211/files#r55738070>.
>



-- 
Stephane Sezer